### PR TITLE
feat: add API key model

### PR DESCRIPTION
**Pull request by stakach**
allows generation of an API key where the key is never saved in the database for security

The idea is that the token generates a decoded JWT token that will be used against the controllers - so only a minor change to the authentication layer is required

### CODE REVIEW

File: src/placeos-models/api_key.cr [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+    attribute secret : String = ->{ Random::Secure.urlsafe_base64(32) }, mass_assignment: false
+
+    belongs_to User
+    belongs_to Authority
+
+    def user=(user)
+      super(user)
+      self.authority_id = user.authority_id
+    end
```
Comment (caspiano) [inline src/placeos-models/api_key.cr:28]: Due to some weirdness with active-model, I think this is the way to do overrides of getters/setters

https://github.com/PlaceOS/models/blob/master/src/placeos-models/authority.cr#L27-L35

File: src/placeos-models/api_key.cr [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+    # Callbacks
+    ###############################################################################################
+
+    before_create :safe_id
+    before_create :set_authority
+    before_create :hash!
+
+    # Reject '+' and '~'
```
Comment (caspiano) [inline src/placeos-models/api_key.cr:63]:
```suggestion
```

File: src/placeos-models/api_key.cr [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    protected def safe_id
+      self._new_flag = true
+      @id ||= Random.new.hex(16)
```
Comment (caspiano) [inline src/placeos-models/api_key.cr:66]: I think this creates a hex string with 16 bytes to entropy, not a string of 16 chars length.

File: src/placeos-models/api_key.cr [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+      # Same error as being unable to find the model
+      if model.secret != OpenSSL::HMAC.hexdigest(:sha512, secret, id)
+        raise RethinkORM::Error::DocumentNotFound.new("Key not present: #{id}")
+      end
+
+      model
+    end
+
+    ISSUER = "POS"
```
Comment (caspiano) [inline src/placeos-models/api_key.cr:102]: Does this constant exist anywhere else?

Comment (stakach) [reply src/placeos-models/api_key.cr:101]: it does not

File: src/placeos-models/api_key.cr (hunk 3 above)
Comment (stakach) [reply src/placeos-models/api_key.cr:65]: nice, good catch

Comment (stakach) [reply src/placeos-models/api_key.cr:65]: ahh yeah so this should be 32 bytes long which seems good

File: src/placeos-models/api_key.cr [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+    # Token Methods
+    ###############################################################################################
+
+    def x_api_key
+      xkey = @x_api_key
+      return xkey if xkey
+      return nil if self.persisted?
+      @x_api_key = "#{self.safe_id}.#{self.secret}"
+    end
```
Comment (caspiano) [inline src/placeos-models/api_key.cr:86]: Is it necessary to prefix this by `x`?

**Review by caspiano — APPROVED** [review]

**Pull request merged**